fdbserver/ptxn/test/FakeLogSystem.h

```diff
@@ -23,92 +23,124 @@
+	                                       int8_t peekLocality,
+	                                       Version localEnd,
+	                                       bool canDiscardPopped) override;
+	virtual Future<Version> getTxsPoppedVersion() override;
+	virtual Version getKnownCommittedVersion() override;
+	virtual Future<Void> onKnownCommittedVersionChange() override;
+	virtual void popTxs(Version upTo, int8_t popLocality) override;
+	virtual void pop(Version upTo, Tag tag, Version knownCommittedVersion, int8_t popLocality) override;
+	virtual Future<Void> confirmEpochLive(Optional<UID> debugID) override;
+	virtual Future<Void> endEpoch() override;
+	virtual Version getEnd() const override;
+	virtual Version getBackupStartVersion() const override;
+	virtual std::map<LogEpoch, EpochTagsVersionsInfo> getOldEpochTagsVersionsInfo() const override;
+	virtual Future<Reference<ILogSystem>> newEpoch(const RecruitFromConfigurationReply& recr,
+	                                               const Future<::RecruitRemoteFromConfigurationReply>& fRemoteWorkers,
+	                                               const DatabaseConfiguration& config,
+	                                               LogEpoch recoveryCount,
+	                                               int8_t primaryLocality,
+	                                               int8_t remoteLocality,
+	                                               const vector<Tag>& allTags,
+	                                               const Reference<AsyncVar<bool>>& recruitmentStalled,
+	                                               TLogGroupCollectionRef tLogGroupCollection) override;
+	virtual LogSystemConfig getLogSystemConfig() const override;
+	virtual Standalone<StringRef> getLogsValue() const override;
+	virtual Future<Void> onLogSystemConfigChange() override;
+	virtual void getPushLocations(VectorRef<Tag> tags, vector<int>& locations, bool allLocations) const override;
+	virtual bool hasRemoteLogs() const override;
+	virtual Tag getRandomRouterTag() const override;
+	virtual int getLogRouterTags() const override;
+	virtual Tag getRandomTxsTag() const override;
+	virtual TLogVersion getTLogVersion() const override;
+	virtual void stopRejoins() override;
+	virtual Tag getPseudoPopTag(Tag tag, ProcessClass::ClassType type) const override;
+	virtual bool hasPseudoLocality(int8_t locality) const override;
+	virtual Version popPseudoLocalityTag(Tag tag, Version upTo) override;
+	virtual void setBackupWorkers(const std::vector<::InitializeBackupReply>& replies) override;
+	virtual bool removeBackupWorker(const BackupWorkerDoneRequest& req) override;
+	virtual LogEpoch getOldestBackupEpoch() const override;
+	virtual void setOldestBackupEpoch(LogEpoch epoch) override;
+};
+
+// FakeLogSystem that returns a FakePeekCursor binding to an id, this allows the cursor can be manipulated
+// outside the scope the LogSystem is created.
+class FakeLogSystem_CustomPeekCursor : public FakeLogSystem {
```

Comment:
Why does a cursor class need to be `ILogSystem` type? And this can be marked as `final`.

Reply:
It returns the CustomPeekCursor, it is an ILogSystem object. Also I intended to mark it not final, for possible inheritance.